File: LONER/AI/loner.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>
#include <bitset>
#include <stack>
#include <algorithm>

using namespace std;

unordered_set<int> solvable_patterns;
const int MAX_CLUSTER = 20;  // precompute up to 20-bit segments

// Precompute solvable patterns for segments of length up to MAX_CLUSTER
void precompute_solvable_patterns(int max_len = MAX_CLUSTER) {
    for (int len = 1; len <= max_len; ++len) {
        int total_states = 1 << len;
        for (int state = 0; state < total_states; ++state) {
            bitset<MAX_CLUSTER> b(state);
            unordered_set<int> visited;
            stack<int> stk;
            stk.push(state);
            bool solvable = false;

            while (!stk.empty()) {
                int s = stk.top(); stk.pop();
                if (visited.count(s)) continue;
                visited.insert(s);

                bitset<MAX_CLUSTER> bs(s);
                if (bs.count() == 1) {
                    solvable = true;
                    break;
                }

                for (int i = 0; i < len - 2; ++i) {
                    if (bs[i] && bs[i + 1] && !bs[i + 2]) {
                        bitset<MAX_CLUSTER> next = bs;
                        next[i] = 0;
                        next[i + 1] = 0;
                        next[i + 2] = 1;
                        stk.push((int)(next.to_ulong()));
                    }
                }

                for (int i = 2; i < len; ++i) {
                    if (!bs[i - 2] && bs[i - 1] && bs[i]) {
                        bitset<MAX_CLUSTER> next = bs;
                        next[i] = 0;
                        next[i - 1] = 0;
                        next[i - 2] = 1;
                        stk.push((int)(next.to_ulong()));
                    }
                }
            }

            if (solvable) {
                solvable_patterns.insert((len << MAX_CLUSTER) | state);  // encode length with state
            }
        }
    }
}

// Check if a cluster is solvable using precomputed patterns
bool is_cluster_solvable(const string& cluster) {
    int len = cluster.length();
    if (len > MAX_CLUSTER) return false;
    int mask = 0;
    for (int i = 0; i < len; ++i) {
        if (cluster[i] == '1') {
            mask |= (1 << i);
        }
    }
    return solvable_patterns.count((len << MAX_CLUSTER) | mask);
}

bool can_win(const string& board) {
    int n = board.length();
    vector<string> clusters;
    string current;
    int zero_run = 2;  // starts at 2 to force a split at the beginning
    int total_pawns = 0;

    for (int i = 0; i <= n; ++i) {
        char c = (i < n) ? board[i] : '0';  // append sentinel 0
        if (c == '1') {
            if (zero_run >= 2 && !current.empty()) {
                // A new cluster starts
                if (!is_cluster_solvable(current)) return false;
                total_pawns += count(current.begin(), current.end(), '1');
                current.clear();
            }
            zero_run = 0;
            current += '1';
        } else {
            zero_run++;
            if (!current.empty()) current += '0';
        }
    }

    if (!current.empty()) {
        if (!is_cluster_solvable(current)) return false;
        total_pawns += count(current.begin(), current.end(), '1');
    }

    return total_pawns == 1;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    precompute_solvable_patterns();  // Precompute patterns up to MAX_CLUSTER size

    int t;
    cin >> t;
    while (t--) {
        int n;
        string board;
        cin >> n >> board;
        cout << (can_win(board) ? "yes" : "no") << '\n';
    }

    return 0;
}

File: LONER/AI/loner.py
def can_win(board_str):
    n = len(board_str)

    # Convert board string to bitmask integer
    def to_mask(s):
        return int(s, 2)

    def count_bits(x):
        return bin(x).count('1')

    def get_bit(mask, i):
        return (mask >> i) & 1

    def set_bit(mask, i, val):
        if val:
            return mask | (1 << i)
        else:
            return mask & ~(1 << i)

    start = to_mask(board_str)
    visited = set()
    stack = [start]

    while stack:
        state = stack.pop()
        if state in visited:
            continue
        visited.add(state)

        if count_bits(state) == 1:
            return "yes"

        for i in range(n):
            # Right jump: 1(i),1(i+1),0(i+2) → 0(i),0(i+1),1(i+2)
            if i + 2 < n and get_bit(state, i) and get_bit(state, i + 1) and not get_bit(state, i + 2):
                new_state = state
                new_state = set_bit(new_state, i, 0)
                new_state = set_bit(new_state, i + 1, 0)
                new_state = set_bit(new_state, i + 2, 1)
                stack.append(new_state)

            # Left jump: 0(i-2),1(i-1),1(i) → 1(i-2),0(i-1),0(i)
            if i - 2 >= 0 and not get_bit(state, i - 2) and get_bit(state, i - 1) and get_bit(state, i):
                new_state = state
                new_state = set_bit(new_state, i, 0)
                new_state = set_bit(new_state, i - 1, 0)
                new_state = set_bit(new_state, i - 2, 1)
                stack.append(new_state)

    return "no"

# Driver
t = int(input())
for _ in range(t):
    n = int(input())
    board = input().strip()
    print(can_win(board))

